packages/widget.vite-plugin.unstable/src/client/app.tsx

```diff
@@ -41,53 +47,66 @@ export const App: React.FC = () => {
   // Poll the finish endpoint until it returns a success or error
   useEffect(() => {
     const poll = window.setInterval(() => {
-      void finish().then((result) => {
-        if (result.status === "success") {
-          setLoading({ state: "success" });
-          setTimeout(() => {
-            window.location.href = result.redirectUrl;
-          }, REDIRECT_DELAY);
+      void finish()
+        .then((result) => {
+          if (result.status === "pending") {
+            return;
+          }
           window.clearInterval(poll);
-        } else if (result.status === "error") {
-          setLoading({ state: "failed", error: result.error });
+          setPageState(
+            result.status === "success"
+              ? { state: "success", redirectUrl: result.redirectUrl }
+              : { state: "failed", error: result.error },
+          );
+        })
+        .catch((error: unknown) => {
           window.clearInterval(poll);
-        }
-      }).catch((error: unknown) => {
-        setLoading({
-          state: "failed",
-          error: error instanceof Error ? error.message : undefined,
+          setPageState({
+            state: "failed",
+            error: error instanceof Error ? error.message : undefined,
+          });
         });
-        window.clearInterval(poll);
-      });
     }, POLLING_INTERVAL);
     return () => window.clearInterval(poll);
   }, []);
 
+  // Redirect on reaching success state
+  useEffect(() => {
+    // When running in Code Workspaces the parent app will handle the redirect
+    if (pageState.state === "success" && pageState.redirectUrl != null) {
+      setTimeout(() => {
+        window.location.href = pageState.redirectUrl!;
+      }, REDIRECT_DELAY);
+    }
+  }, [pageState]);
```

Comment:
Should we still kick off the redirect once we set the page state rather than doing it separately in a useEffect? The rendering of the text will still work based on the set page state but it might be cleaner to do the redirect in the existing place rather than doing it afterwards in a side effect